TestWatchReconnRequest fails

```
--- FAIL: TestWatchReconnRequest (5.35s)
	watch_test.go:281: watch timed out
```

Another one

```
=== RUN   TestWatchReconnRequest
2016-02-12 18:08:32.634256 I | etcdserver: name = node514218490942812466
2016-02-12 18:08:32.634281 I | etcdserver: data dir = /tmp/etcd691073929
2016-02-12 18:08:32.634328 I | etcdserver: member dir = /tmp/etcd691073929/member
2016-02-12 18:08:32.634339 I | etcdserver: heartbeat = 10ms
2016-02-12 18:08:32.634348 I | etcdserver: election = 100ms
2016-02-12 18:08:32.634357 I | etcdserver: snapshot count = 0
2016-02-12 18:08:32.634373 I | etcdserver: advertise client URLs = http://127.0.0.1:20408
2016-02-12 18:08:32.634385 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:20407
2016-02-12 18:08:32.634409 I | etcdserver: initial cluster = node329837469360592218=http://127.0.0.1:20405,node4785672846810766913=http://127.0.0.1:20403,node514218490942812466=http://127.0.0.1:20407
2016-02-12 18:08:32.634450 I | etcdserver: name = node4785672846810766913
2016-02-12 18:08:32.634467 I | etcdserver: data dir = /tmp/etcd375594751
2016-02-12 18:08:32.634476 I | etcdserver: member dir = /tmp/etcd375594751/member
2016-02-12 18:08:32.634482 I | etcdserver: heartbeat = 10ms
2016-02-12 18:08:32.634487 I | etcdserver: election = 100ms
2016-02-12 18:08:32.634492 I | etcdserver: snapshot count = 0
2016-02-12 18:08:32.634500 I | etcdserver: advertise client URLs = http://127.0.0.1:20404
2016-02-12 18:08:32.634506 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:20403
2016-02-12 18:08:32.634519 I | etcdserver: initial cluster = node329837469360592218=http://127.0.0.1:20405,node4785672846810766913=http://127.0.0.1:20403,node514218490942812466=http://127.0.0.1:20407
2016-02-12 18:08:32.634720 I | etcdserver: name = node329837469360592218
2016-02-12 18:08:32.634752 I | etcdserver: data dir = /tmp/etcd750839378
2016-02-12 18:08:32.634763 I | etcdserver: member dir = /tmp/etcd750839378/member
2016-02-12 18:08:32.634772 I | etcdserver: heartbeat = 10ms
2016-02-12 18:08:32.634780 I | etcdserver: election = 100ms
2016-02-12 18:08:32.634787 I | etcdserver: snapshot count = 0
2016-02-12 18:08:32.634796 I | etcdserver: starting member 2211da3935471694 in cluster 47111e9824bc06a8
2016-02-12 18:08:32.634826 I | raft: 2211da3935471694 became follower at term 0
2016-02-12 18:08:32.634836 I | raft: newRaft 2211da3935471694 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-02-12 18:08:32.634841 I | raft: 2211da3935471694 became follower at term 1
2016-02-12 18:08:32.634921 I | etcdserver: advertise client URLs = http://127.0.0.1:20406
2016-02-12 18:08:32.634942 I | etcdserver: initial advertise peer URLs = http://127.0.0.1:20405
2016-02-12 18:08:32.634958 I | etcdserver: initial cluster = node329837469360592218=http://127.0.0.1:20405,node4785672846810766913=http://127.0.0.1:20403,node514218490942812466=http://127.0.0.1:20407
2016-02-12 18:08:32.635110 I | etcdserver: starting member 1acd4738a081feb4 in cluster 47111e9824bc06a8
2016-02-12 18:08:32.635165 I | raft: 1acd4738a081feb4 became follower at term 0
2016-02-12 18:08:32.635186 I | raft: newRaft 1acd4738a081feb4 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-02-12 18:08:32.635199 I | raft: 1acd4738a081feb4 became follower at term 1
2016-02-12 18:08:32.635286 I | etcdserver: starting member ca90a3d83571719 in cluster 47111e9824bc06a8
2016-02-12 18:08:32.635313 I | raft: ca90a3d83571719 became follower at term 0
2016-02-12 18:08:32.635323 I | raft: newRaft ca90a3d83571719 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2016-02-12 18:08:32.635329 I | raft: ca90a3d83571719 became follower at term 1
2016-02-12 18:08:32.648948 I | etcdserver: set snapshot count to default 10000
2016-02-12 18:08:32.648998 I | etcdserver: starting server... [version: 2.3.0-alpha.0+git, cluster version: to_be_decided]
2016-02-12 18:08:32.649364 I | etcdserver: set snapshot count to default 10000
2016-02-12 18:08:32.649391 I | etcdserver: starting server... [version: 2.3.0-alpha.0+git, cluster version: to_be_decided]
2016-02-12 18:08:32.650306 I | etcdserver: set snapshot count to default 10000
2016-02-12 18:08:32.650333 I | etcdserver: starting server... [version: 2.3.0-alpha.0+git, cluster version: to_be_decided]
2016-02-12 18:08:32.650750 N | etcdserver: added member ca90a3d83571719 [http://127.0.0.1:20405] to cluster 47111e9824bc06a8
2016-02-12 18:08:32.650837 N | etcdserver: added member 1acd4738a081feb4 [http://127.0.0.1:20407] to cluster 47111e9824bc06a8
2016-02-12 18:08:32.650919 N | etcdserver: added local member 2211da3935471694 [http://127.0.0.1:20403] to cluster 47111e9824bc06a8
2016-02-12 18:08:32.651131 N | etcdserver: added member ca90a3d83571719 [http://127.0.0.1:20405] to cluster 47111e9824bc06a8
2016-02-12 18:08:32.651268 N | etcdserver: added local member 1acd4738a081feb4 [http://127.0.0.1:20407] to cluster 47111e9824bc06a8
2016-02-12 18:08:32.651362 N | etcdserver: added member 2211da3935471694 [http://127.0.0.1:20403] to cluster 47111e9824bc06a8
2016-02-12 18:08:32.651412 I | rafthttp: the connection with 2211da3935471694 became active
2016-02-12 18:08:32.651473 I | rafthttp: the connection with ca90a3d83571719 became active
2016-02-12 18:08:32.651541 N | etcdserver: added local member ca90a3d83571719 [http://127.0.0.1:20405] to cluster 47111e9824bc06a8
2016-02-12 18:08:32.651589 N | etcdserver: added member 1acd4738a081feb4 [http://127.0.0.1:20407] to cluster 47111e9824bc06a8
2016-02-12 18:08:32.651611 I | rafthttp: the connection with 1acd4738a081feb4 became active
2016-02-12 18:08:32.651638 N | etcdserver: added member 2211da3935471694 [http://127.0.0.1:20403] to cluster 47111e9824bc06a8
2016-02-12 18:08:32.651706 I | rafthttp: the connection with ca90a3d83571719 became active
2016-02-12 18:08:32.652277 I | rafthttp: the connection with 2211da3935471694 became active
2016-02-12 18:08:32.652563 I | rafthttp: the connection with 1acd4738a081feb4 became active
2016-02-12 18:08:32.675392 I | raft: 1acd4738a081feb4 is starting a new election at term 1
2016-02-12 18:08:32.675457 I | raft: 1acd4738a081feb4 became candidate at term 2
2016-02-12 18:08:32.675464 I | raft: 1acd4738a081feb4 received vote from 1acd4738a081feb4 at term 2
2016-02-12 18:08:32.675476 I | raft: 1acd4738a081feb4 [logterm: 1, index: 3] sent vote request to ca90a3d83571719 at term 2
2016-02-12 18:08:32.675487 I | raft: 1acd4738a081feb4 [logterm: 1, index: 3] sent vote request to 2211da3935471694 at term 2
2016-02-12 18:08:32.675639 I | raft: ca90a3d83571719 [term: 1] received a MsgVote message with higher term from 1acd4738a081feb4 [term: 2]
2016-02-12 18:08:32.675654 I | raft: ca90a3d83571719 became follower at term 2
2016-02-12 18:08:32.675673 I | raft: ca90a3d83571719 [logterm: 1, index: 3, vote: 0] voted for 1acd4738a081feb4 [logterm: 1, index: 3] at term 2
2016-02-12 18:08:32.675694 I | raft: 2211da3935471694 [term: 1] received a MsgVote message with higher term from 1acd4738a081feb4 [term: 2]
2016-02-12 18:08:32.675714 I | raft: 2211da3935471694 became follower at term 2
2016-02-12 18:08:32.675738 I | raft: 2211da3935471694 [logterm: 1, index: 3, vote: 0] voted for 1acd4738a081feb4 [logterm: 1, index: 3] at term 2
2016-02-12 18:08:32.675770 I | raft: 1acd4738a081feb4 received vote from ca90a3d83571719 at term 2
2016-02-12 18:08:32.675777 I | raft: 1acd4738a081feb4 [quorum:2] has received 2 votes and 0 vote rejections
2016-02-12 18:08:32.675789 I | raft: 1acd4738a081feb4 became leader at term 2
2016-02-12 18:08:32.675798 I | raft: raft.node: 1acd4738a081feb4 elected leader 1acd4738a081feb4 at term 2
2016-02-12 18:08:32.675976 I | raft: raft.node: 2211da3935471694 elected leader 1acd4738a081feb4 at term 2
2016-02-12 18:08:32.676000 I | raft: raft.node: ca90a3d83571719 elected leader 1acd4738a081feb4 at term 2
2016-02-12 18:08:32.676459 I | etcdserver: published {Name:node514218490942812466 ClientURLs:[http://127.0.0.1:20408]} to cluster 47111e9824bc06a8
2016-02-12 18:08:32.676604 I | etcdserver: published {Name:node329837469360592218 ClientURLs:[http://127.0.0.1:20406]} to cluster 47111e9824bc06a8
2016-02-12 18:08:32.676750 I | etcdserver: published {Name:node4785672846810766913 ClientURLs:[http://127.0.0.1:20404]} to cluster 47111e9824bc06a8
2016-02-12 18:08:32.676866 I | etcdserver: setting up the initial cluster version to 2.3
2016-02-12 18:08:32.677062 N | etcdserver: set the initial cluster version to 2.3
2016-02-12 18:08:32.677146 N | etcdserver: set the initial cluster version to 2.3
2016-02-12 18:08:32.677182 N | etcdserver: set the initial cluster version to 2.3
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016/02/12 18:08:32 grpc: Server.Serve failed to create ServerTransport:  connection error: desc = "transport: write unix localhost:node4785672846810766913.sock->@: write: broken pipe"
2016-02-12 18:08:32.817192 E | rafthttp: failed to read ca90a3d83571719 on stream Message (net/http: request canceled)
2016-02-12 18:08:32.817225 I | rafthttp: the connection with ca90a3d83571719 became inactive
2016-02-12 18:08:32.817715 E | rafthttp: failed to read 1acd4738a081feb4 on stream MsgApp v2 (net/http: request canceled)
2016-02-12 18:08:32.817733 I | rafthttp: the connection with 1acd4738a081feb4 became inactive
2016-02-12 18:08:32.818638 E | rafthttp: failed to read 1acd4738a081feb4 on stream MsgApp v2 (net/http: request canceled)
2016-02-12 18:08:32.818657 I | rafthttp: the connection with 1acd4738a081feb4 became inactive
--- FAIL: TestWatchReconnRequest (0.21s)
	cluster.go:723: grpc: the client connection is closing
```

--- a/clientv3/integration/watch_test.go
+++ b/clientv3/integration/watch_test.go
@@ -163,13 +163,18 @@ func TestWatchReconnRequest(t *testing.T) {
 }
 
 func testWatchReconnRequest(t *testing.T, wctx *watchctx) {
-	// take down watcher connection
-	donec := make(chan struct{})
+	donec, stopc := make(chan struct{}), make(chan struct{}, 1)
 	go func() {
+		timer := time.After(2 * time.Second)
+		defer close(donec)
+		// take down watcher connection
 		for {
 			wctx.wclient.ActiveConnection().Close()
 			select {
-			case <-donec:
+			case <-timer:
+				// spinning on close may live lock reconnection
+				return
+			case <-stopc:
 				return
 			default:
 			}
@@ -179,7 +184,11 @@ func testWatchReconnRequest(t *testing.T, wctx *watchctx) {
 	if wctx.ch = wctx.w.Watch(context.TODO(), "a", 0); wctx.ch == nil {
 		t.Fatalf("expected non-nil channel")
 	}
-	close(donec)
+
+	// wait for disconnections to stop
+	stopc <- struct{}{}
+	<-donec
+
 	// ensure watcher works
 	putAndWatch(t, wctx, "a", "a")
 }

--- a/clientv3/watch.go
+++ b/clientv3/watch.go
@@ -198,6 +198,12 @@ func (w *watcher) addStream(resp *pb.WatchResponse, pendingReq *watchRequest) {
 		resumec: make(chan int64),
 	}
 
+	if pendingReq.rev == 0 {
+		// note the header revision so that a put following a current watcher
+		// disconnect will arrive on the watcher channel after reconnect
+		ws.initReq.rev = resp.Header.Revision
+	}
+
 	w.mu.Lock()
 	w.streams[ws.id] = ws
 	w.mu.Unlock()

